WeakAuras/RegionTypes/DynamicGroup.lua

```diff
@@ -813,7 +814,46 @@ local function modify(parent, region, data)
     WeakAuras.StartProfileSystem("dynamicgroup")
     WeakAuras.StartProfileAura(data.id)
     local newPositions = {}
-    self.growFunc(newPositions, self.sortedChildren)
+    local anchorPerUnit = data.anchorFrameType == "NAMEPLATE" or data.anchorFrameType == "UNITFRAME"
+    if data and anchorPerUnit then
+      local units = {}
+      -- build a list keyed by unit
+      for i, regionData in ipairs(self.sortedChildren) do
+        local anchor, unit
+        local state = regionData.region.state
+        if state and state.unit then
+          unit = state.unit
+          if data.anchorFrameType == "NAMEPLATE" then
+            anchor = C_NamePlate.GetNamePlateForUnit(unit)
+          elseif data.anchorFrameType == "UNITFRAME" then
+            anchor = GetUnitFrame(unit)
+          end
+        end
+        if anchor then
+          regionData.region.anchor = anchor
+          units[unit] = units[unit] or {}
+          tinsert(units[unit], i)
+        elseif data.anchorShowUnitsWithNoFrame then
+          units[""] = units[""] or {}
+          tinsert(units[""], i)
+          regionData.region.anchor = nil
+        end
+      end
+      -- run growFunc for each unit
+      for unit, indexes in pairs(units) do
+        local unitActiveRegions = {}
+        local unitNewPositions = {}
+        for _, index in ipairs(indexes) do
+            tinsert(unitActiveRegions, self.sortedChildren[index])
+        end
+        self.growFunc(unitNewPositions, unitActiveRegions)
+        for i, index in ipairs(indexes) do
+          newPositions[index] = unitNewPositions[i]
+        end
+      end
+    else
+      self.growFunc(newPositions, self.sortedChildren)
+    end
     for index, regionData in ipairs(self.sortedChildren) do
```

Comment:
If you extract the whole for loop into a function that takes:
sortedChildren + positions, then you don't need to copy from unitNewPositions to newPositions.

Instead you just call that function per unit.